### PR TITLE
test: add property test coverage to Read Buffer transcoders

### PR DESCRIPTION
**Pull request by e-dard**
This PR adds some quickcheck/property tests to the Read Buffer using the `proptest` crate. 

There is already good test coverage for boundary/edge cases in the Read Buffer, but this add another approach, which I think will be helpful as more transcoders get added.

The general idea is to trust that the basic column encoding (effectively a vector of logical values stored in the same physical type) is correct. Then, we can use property tests to generate test cases for us that we feed into the control encoding and also into other encodings. We then verify that the results of various read filter operations agree.

### CODE REVIEW

File: read_buffer/src/column/encoding/scalar/fixed_null.rs [hunk 1]
```diff
@@ -516,14 +516,17 @@ mod test {
...
+    use super::super::transcoders::{
+        ByteTrimmer, FloatByteTrimmer, MockTranscoder, NoOpTranscoder,
+    };
     use super::cmp::Operator;
     use super::*;
 
     // Helper function to create a new FixedNull encoding using a mock transcoder
     // that will allow tests to track calls to encode/decode.
-    fn new_encoding(
+    fn new_mock_encoding(
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed_null.rs:529]: Just a rename for clarify

File: read_buffer/src/column/encoding/scalar/fixed_null.rs [hunk 2]
```diff
@@ -884,28 +887,241 @@ mod test {
...
+            proptest! {
+                #[test]
+                 // The proptest strategy will generate vectors of values within the physical type
+                 // bounds, ensuring they can be safely encoded.
+                 // The strategy effectively says:
+                 //
+                 //   Generate vectors of Option<T> where the value will be `None`.
+                 //   Generate values according to the provided range, and generate
+                 //   `n` of them according to the size range `0..=50`.
+                fn $fn_name(arr in prop::collection::vec(proptest::option::weighted(0.9, <$physical>::MIN as $logical ..=<$physical>::MAX as $logical), 0..=50)) {
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed_null.rs:930]: Effectively - generate a vector of values within a range specified by the physical type of the column encoding. Replace each value with `None` with 10% probability.

File: read_buffer/src/column/encoding/scalar/rle.rs [hunk 3]
```diff
@@ -1159,4 +1164,116 @@ mod test {
...
+                 // The proptest strategy will generate vectors of values within the physical type
+                 // bounds, ensuring they can be safely encoded.
+                 // The strategy effectively says:
+                 //
+                 //   Generate vectors of Option<T> where the value will be `None`.
+                 //   Generate values according to the provided range, and generate
+                 //   `n` of them according to the size range `0..=5000`. The
+                 //   parameters are set in a way that should exercise the core
+                 //   RLE functionality.
+                fn $fn_name(mut arr in prop::collection::vec(proptest::option::weighted(0.9, 0 as $logical ..=50 as $logical), 0..=1000)) {
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/rle.rs:1186]: For RLE we want to exercise the compression by generating lower cardinality inputs.

File: read_buffer/src/column/encoding/scalar/fixed_null.rs [hunk 4]
```diff
@@ -884,28 +887,241 @@ mod test {
...
         assert!(v.has_any_non_null_value());
     }
+
+    // This macro builds out property tests for the integer byte trimmer encoder.
+    // Each of the supported logical types (i64, u64) is tested with transcoders
+    // that store encoded values physically as (i32, u32, i16, u16, i8, u8)
+    // depending on logical type and value range.
+    macro_rules! make_test_transcoder_integer_bytetrimmer {
+        (($logical:ty, $logical_arrow:ty, $physical:ty, $physical_arrow:ty, $fn_name:ident)) => {
+            proptest! {
```
[the remark this answers is not in the document]
Comment (alamb) [reply read_buffer/src/column/encoding/scalar/fixed_null.rs:921]: macros!

File: read_buffer/src/column/encoding/scalar/rle.rs [hunk 5]
```diff
@@ -1159,4 +1164,116 @@ mod test {
...
+                #[test]
+                 // The proptest strategy will generate vectors of values within the physical type
+                 // bounds, ensuring they can be safely encoded.
+                 // The strategy effectively says:
+                 //
+                 //   Generate vectors of Option<T> where the value will be `None`.
+                 //   Generate values according to the provided range, and generate
+                 //   `n` of them according to the size range `0..=5000`. The
+                 //   parameters are set in a way that should exercise the core
+                 //   RLE functionality.
```
Comment (domodwyer) [inline read_buffer/src/column/encoding/scalar/rle.rs:1185]: The proptest strategies are quite dense once composed up - comments like this make it much nicer to read through 👍

File: read_buffer/Cargo.toml [hunk 6]
```diff
@@ -30,6 +30,7 @@ workspace-hack = { path = "../workspace-hack"}
 [dev-dependencies] # In alphabetical order
 criterion = "0.3.3"
 packers = { path = "../packers" }
+proptest = "1.0.0"
```
Comment (alamb) [inline read_buffer/Cargo.toml:33]: TIL https://github.com/AltSysrq/proptest#introduction

File: read_buffer/src/column/encoding/scalar/fixed_null.rs [hunk 7]
```diff
@@ -884,28 +887,241 @@ mod test {
...
+        };
+    }
+
+    make_test_transcoder_integer_bytetrimmer!((
+        u64,
+        UInt64Type,
+        u8,
+        UInt8Type,
+        test_transcoder_byte_trim_u64_to_u8
+    ));
```
Comment (domodwyer) [inline read_buffer/src/column/encoding/scalar/fixed_null.rs:987]: I love how clearly these test cases are expressed using macros - it would have been painful if they were all nearly-identical functions ❤️

**Review by domodwyer — APPROVED** [review]
No code fixes I see 💅 

Very nice stuff - I love prop testing, and I love the macro approach.

File: read_buffer/src/column/encoding/scalar/fixed_null.rs (hunk 2 above)
Comment (alamb) [inline read_buffer/src/column/encoding/scalar/fixed_null.rs:930]:
```suggestion
               // Effectively - generate a vector of values within a range specified by the 
               // physical type of the column encoding. Replace each value with None 
               // with 10% probability.
                fn $fn_name(arr in prop::collection::vec(proptest::option::weighted(0.9, <$physical>::MIN as $logical ..=<$physical>::MAX as $logical), 0..=50)) {
```

File: read_buffer/src/column/encoding/scalar/fixed_null.rs [hunk 8]
```diff
@@ -884,28 +887,241 @@ mod test {
...
+                            .map(|v| v.map(|v| transcoder.encode(v)))
+                            .collect::<PrimitiveArray<_>>(), // encode u64 as u8,
+                        transcoder,
+                    );
+
+                    // exercise some physical operations
+                    let mut cases = vec![];
+                    for op in ["<", "<=", ">", ">=", "=", "!="] {
+                        for v in [
+                            <$physical>::MIN,
```
Comment (alamb) [inline read_buffer/src/column/encoding/scalar/fixed_null.rs:951]: `$physical` in this case is the value generated by proptest, right?

**Review by alamb — APPROVED** [review]
This is very cool -- I went through it fairly carefully and it looks 👨‍🍳  to me
